src/harness/reference_models/ppa_model/ppa_ref_model.py

```diff
@@ -0,0 +1,82 @@
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from concurrent.futures import ProcessPoolExecutor as Pool
+import multiprocessing
+from utils import *
+from compute import load_partials
+from reference_models.geo.vincenty import GeodesicPoint
+
+THRESHOLD = -96
+
+
+def __compute_points_gain_prop(device):
+  pool = Pool(multiprocessing.cpu_count())
+  install_param = device['installationParam']
+  # Get all the Function Partials
+  great_circle_partial, antenna_gain_partial, \
+  prop_partial, cnt_partial = load_partials(install_param, THRESHOLD)
+  # Compute all the Points in 0-359 every 200m upto 40km
+  points = pool.map(great_circle_partial, range(0, 360))
+  # Compute the Gain for all Direction
+  antenna_gain = pool.map(antenna_gain_partial, range(0, 360))
+  # Compute the Path Loss
+  path_loss = pool.map(prop_partial, enumerate(points))
+  # Compute the Contour based on Gain and Path Loss Comparing with Threshold
+  cnt = pool.map(cnt_partial, zip(antenna_gain, path_loss))
+  # Smoothing Contour using Hamming Filter
+  cnt = smooth(np.asarray(list(cnt)))
+  print len(cnt)
+  # Generating lat, lon for Contours
+  lat = []
+  lon = []
+  for c, az in zip(cnt, range(0, 360)):
+    l, lo, az = GeodesicPoint(install_param['latitude'],
+                                 install_param['longitude'],
+                                 c, float(az))
+    lat.append(l)
+    lon.append(lo)
+  polygon = geojson.Polygon([zip(lon,lat)])
+  return polygon
+
+
+def ppa_creation_model(device_filenames, pal_record_filenames,
+                       pal_user_id, pal_low_frequency, pal_high_frequency):
+  """Create a PPA Polygon based on the PAL Records and Device Information
+  Args:
+    device_filenames: (List) A list containing device filenames.
+    pal_record_filenames: (List) A list containing pal record filenames.
+    pal_user_id: (String) Pal User Id to be used in all Pal Records. 
+    pal_low_frequency: (Number) Pal Primary Low Frequency to be assigned 
+    to all Pal Records. 
+    pal_high_frequency: (Number) Pal Primary High Frequency to be assigned 
+    to all Pal Records. 
+  Returns:
+     A Tuple containing PPA Polygon in GeoJSON, a List of 
+     Consistent PAL Record and Device List
+   Note: Device Records must contain eirpCapability in installationParam Object
+   and Pal Records must contain fipsCode
+  """
+  devices, pal_records = load_files(device_filenames, pal_record_filenames,
+                                   pal_user_id, pal_low_frequency, pal_high_frequency)
+  pool = Pool(multiprocessing.cpu_count())
```

Comment:
I would keep the usage of the pool only here, where you parallelize per device. Doing further fine grain parallelization inside __compute_points_gain() should not give any additional speed gain (and actually probably some slow down due to contention and context switch), provided that number of CBSD  > number of CPU.
But in any case if only small number of CBSD, the calculation will be fast enough so no real need of fancy speed up.